lib/puppet/indirector/node/rest.rb

```diff
@@ -4,4 +4,14 @@
 class Puppet::Node::Rest < Puppet::Indirector::REST
   desc "Get a node via REST. Puppet agent uses this to allow the puppet master
     to override its environment."
+
+  def find(request)
+    super(request) do |response|
+      server_version = response['X-Puppet-Version']
+      if SemanticPuppet::Version.parse(server_version).major < SemanticPuppet::Version.parse(Puppet::PUPPETVERSION).major
+        Puppet.warning "Server version #{server_version} is unsupported, downgrading to PSON"
+        Puppet[:preferred_serialization_format] = 'pson'
```

Comment:
It feels odd to me to have this in the node find - seems to rely on the node find being the first thing that happens during the agent run and it only happening once, which might not always be the case (isn't a guaranteed contract?). Interested to hear why you selected this location. Pros/cons of placing in `Puppet::Indirector::Rest` with some guard against re-evaluating more than once?